src/middleware.ts

```diff
@@ -259,43 +296,59 @@ export const persist = <S extends State>(
   }
 
   // rehydrate initial state with existing stored state
-  ;(async () => {
-    const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
-
-    try {
-      const storageValue = await storage.getItem(name)
 
+  // a workaround to solve the issue of not storing rehydrated state in sync storage
+  // the set(state) value would be later overridden with initial state by create()
+  // to avoid this, we merge the state from localStorage into the initial state.
+  // TODO: find a better solution for this
+  let stateFromStorage: any
+  const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
+  // bind is used to avoid `TypeError: Illegal invocation` error
+  toThenable(storage.getItem.bind(storage))(name)
+    .then((storageValue) => {
       if (storageValue) {
-        const deserializedStorageValue = await deserialize(storageValue)
-
-        // if versions mismatch, run migration
+        return deserialize(storageValue)
+      }
+    })
+    .then((deserializedStorageValue) => {
+      if (deserializedStorageValue) {
         if (deserializedStorageValue.version !== version) {
-          const migratedState = await migrate?.(
-            deserializedStorageValue.state,
-            deserializedStorageValue.version
-          )
-          if (migratedState) {
-            set(migratedState)
-            await setItem()
+          if (migrate) {
+            stateFromStorage = migrate(
+              deserializedStorageValue.state,
+              deserializedStorageValue.version
+            )
+            return stateFromStorage
           }
         } else {
+          stateFromStorage = deserializedStorageValue.state
           set(deserializedStorageValue.state)
         }
       }
-    } catch (e) {
+    })
```

Comment:
Not sure if it'd work, but to solve the above `todo`: maybe we could return instead of assigning to `stateFromStorage`, do a little condition to return the final state, or else, continue the `.then` chain after that?

Reply:
returning some value from inside the then is a little challenging. specially it's a combination of sync/async `then` chain.

Reply:
The goal of assigning the state to `stateFromStorage` is to have the value available outside the closures of promise callbacks. This way it can be returned as initial state: https://github.com/pmndrs/zustand/pull/403/files#diff-0626369278c89781505d2c04865ecce1302e979271bcfb471e029c19addc13a0R352.

Returning it from within a `then` callback would just get the value wrapped into another `thenable`, it wouldn't be returned to the `persist` function. Right?

Reply:
yeah, it will also make the `toThenable()` function non general

Reply:
Ok I see the issue now. Indeed, it wouldn't solve it :confused: